{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA SEQUENCE COMPARISON\n"
     ]
    }
   ],
   "source": [
    "#This codes can be used to compare 2 DNA sequences, transcribe DNA to RNA,and translate RNA to Protein sequence.\n",
    "\n",
    "#This part codes for the DNA comparision: It will display the position of difference and the difference in nucleotides.\n",
    "\n",
    "def sequence_comparision(a,b):\n",
    "    similar=\"\"\n",
    "    extra=\"\"\n",
    "    print(\"total base pair in 1st dna sequence=\",len(a))\n",
    "    print(\"total base pair in 2nd dna sequence=\",len(b))\n",
    "    \n",
    "    temp_a=a.upper()\n",
    "    temp_b=b.upper()\n",
    "    count=0\n",
    "    if len(temp_a)>len(temp_b):\n",
    "        diff=len(temp_a)-len(temp_b)\n",
    "        for i in range(0,len(temp_b)):\n",
    "            if temp_a[i]==temp_b[i]:\n",
    "                similar+=temp_b[i]\n",
    "                continue\n",
    "            else:\n",
    "                count+=1\n",
    "                print('Differnece of base pair at',(i+1),'position.', '\\nBasepair in sequence 1=',temp_a[i],'\\nBasepair in sequence 2=',temp_b[i])\n",
    "        for ex in range (len(temp_b),len(temp_a)):\n",
    "            extra+=temp_a[ex]\n",
    "        print(\"Extra sequence=\", extra)\n",
    "    else:\n",
    "        for j in range (0,len(temp_a)):\n",
    "            if temp_a[j]==temp_b[j]:\n",
    "                similar+=temp_a[j]\n",
    "                continue\n",
    "            else:\n",
    "                count+=1\n",
    "                print('Differnece of base pair at',(j+1),'position.', '\\nBasepair in sequence 1=',temp_a[j],'\\nBasepair in sequence 2=',temp_b[j])\n",
    "        for ex in range (len(temp_a),len(temp_b)):\n",
    "            extra+=temp_b[ex]\n",
    "        print(\"Extra sequence=\", extra)\n",
    "    print(\"number of differences=\", count)\n",
    "    print(\"similar sequence=\",similar)\n",
    "print(\"DNA SEQUENCE COMPARISON\")\n",
    "a_name= input(\"Enter name of sequence 1:\")    \n",
    "a= input(\"\\nEnter sequence here:\")\n",
    "a_name= input(\"\\nEnter name of sequence 2:\") \n",
    "b= input(\"\\nEnter sequence here:\")\n",
    "sequence_comparision(a,b)\n",
    "\n",
    "#This codes to extract the exons\n",
    "exon_a=\"\"\n",
    "for chr in a:\n",
    "    if chr==\"a\"or chr==\"t\" or chr==\"g\" or chr==\"c\":\n",
    "        continue\n",
    "    else:\n",
    "        exon_a+=chr\n",
    "\n",
    "exon_b=\"\"\n",
    "for chrt in b:\n",
    "    if chrt==\"a\"or chrt==\"t\" or chrt==\"g\" or chrt==\"c\":\n",
    "        continue\n",
    "    else:\n",
    "        exon_b+=chrt\n",
    "\n",
    "\n",
    "#This function will calculate the frequency of the 4 nucleotides in both the sequences and then calculate the PCR tempurate for both the sequences.\n",
    "def percentage_basepairs(sequence):\n",
    "    total_base_pairs= len(sequence)\n",
    "    adenine=0\n",
    "    guanine=0\n",
    "    thymine=0\n",
    "    cytosine=0\n",
    "    for i in sequence:\n",
    "        if i=='A'or i==\"a\":\n",
    "            adenine+=1\n",
    "        if i=='T'or i==\"t\":\n",
    "            thymine+=1\n",
    "        if i=='G'or i==\"g\":\n",
    "            guanine+=1\n",
    "        if i=='C'or i==\"c\":\n",
    "            cytosine+=1\n",
    "    per_a=(adenine/total_base_pairs)*100\n",
    "    per_t=(thymine/total_base_pairs)*100\n",
    "    per_g=(guanine/total_base_pairs)*100\n",
    "    per_c=(cytosine/total_base_pairs)*100\n",
    "    print(\"Percentage of A=\",per_a,\"\\nPercentage of T=\",per_t,\"\\nPercentage of G=\",per_g,\"\\nPercentage of C=\",per_c)\n",
    "    return per_a, per_t, per_g, per_c\n",
    "print(\"Nucleotide percentage of sequence 1:\\n\",percentage_basepairs(a))\n",
    "print(\"Nucleotide percentage of sequence 2:\\n\",percentage_basepairs(b))\n",
    "\n",
    "#This function is defined to trascribe DNA sequnces to RNA sequences and RNA sequence to Protein sequence.\n",
    "def transcription_translation(dna):\n",
    "    for ch in dna:\n",
    "        rna=dna.replace(\"T\",\"U\")\n",
    "    print(\"\\nRNA sequence:\",rna)\n",
    "    \n",
    "    #This dictionary contain the triplet codes as keys and their corresponding amino acids as their value\n",
    "    codon={ \"AAA\":\"Lys\", \"AAC\":\"Asn\", \"AAG\":\"Lys\", \"AAU\":\"Asn\", \n",
    "            \"ACA\":\"Thr\", \"ACC\":\"Thr\", \"ACG\":\"Thr\", \"ACU\":\"Thr\", \n",
    "            \"AGA\":\"Arg\", \"AGC\":\"Ser\", \"AGG\":\"Arg\", \"AGU\":\"Ser\", \n",
    "            \"AUA\":\"Ile\", \"AUC\":\"Ile\", \"AUG\":\"Met\", \"AUU\":\"Ile\", \n",
    "\n",
    "            \"CAA\":\"Gln\", \"CAC\":\"His\", \"CAG\":\"Gln\", \"CAU\":\"His\", \n",
    "            \"CCA\":\"Pro\", \"CCC\":\"Pro\", \"CCG\":\"Pro\", \"CCU\":\"Pro\", \n",
    "            \"CGA\":\"Arg\", \"CGC\":\"Arg\", \"CGG\":\"Arg\", \"CGU\":\"Arg\", \n",
    "            \"CUA\":\"Lys\", \"CUC\":\"Leu\", \"CUG\":\"Leu\", \"CUU\":\"Leu\", \n",
    "\n",
    "            \"GAA\":\"Glu\", \"GAC\":\"Asp\", \"GAG\":\"Glu\", \"GAU\":\"Asp\", \n",
    "            \"GCA\":\"Ala\", \"GCC\":\"Ala\", \"GCG\":\"Ala\", \"GCU\":\"Ala\", \n",
    "            \"GGA\":\"Gly\", \"GGC\":\"Gly\", \"GGG\":\"Gly\", \"GGU\":\"Gly\", \n",
    "            \"GUA\":\"Val\", \"GUC\":\"Val\", \"GUG\":\"Val\", \"GUU\":\"Val\", \n",
    "\n",
    "            \"UAA\":\"_\", \"UAC\":\"Tyr\", \"UAG\":\"_\", \"UAU\":\"Thr\", \n",
    "            \"UCA\":\"Ser\", \"UCC\":\"Ser\", \"UCG\":\"Ser\", \"UCU\":\"Ser\", \n",
    "            \"UGA\":\"_\", \"UGC\":\"Cys\", \"UGG\":\"Trp\", \"UGU\":\"Cys\", \n",
    "            \"UUA\":\"Leu\", \"UUC\":\"Phe\", \"UUG\":\"Leu\", \"UUU\":\"Phe\"}\n",
    "    word=[rna[i:i+3] for i in range(0, len(rna), 3)] \n",
    "    print(\"\\nCodon in RNA: \",word)\n",
    "    protein=\"\"\n",
    "    for triplet in word:\n",
    "        if triplet in codon:\n",
    "            protein=protein+codon[triplet]+'-'\n",
    "    print(\"\\nProtein sequence:\",protein)\n",
    "print(\"\\n*******************************\")\n",
    "print(\"\\nExon for sequence 1=\",exon_a)\n",
    "transcription_translation(exon_a)\n",
    "print(\"\\n*******************************\")\n",
    "print(\"\\nExon for sequence 2=\",exon_b)\n",
    "transcription_translation(exon_b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
